Tests for tabs/tabs-1/tabs.html

spectranaut opened the pull request.

spectranaut commented:
@sh0ji -- this is the week Jesse is out, will you be reviewing, or was it @tatermelon?

tatermelon commented on test/util/assertAriaRole.js:
```diff
@@ -0,0 +1,36 @@
+ * @param {String} elementTag - the element the role should be found on
+ */
+
+module.exports = async function assertAriaRole (t, exampleId, role, roleCount, elementTag) {
```
Maybe rename this to assertAriaRoles since you are checking for (possibly) multiple instances of a role

spectranaut commented on test/util/assertAriaRole.js:
```diff
@@ -0,0 +1,36 @@
+    assert.equal(
+      await element.getTagName(),
+      elementTag,
+      'role="' + role + '" should be found on "' + elementTag + '" elements only'
```
In all previous tests I have been testing whether the role is found on the element listed in the attribute table as well. You can see in this [example](https://w3c.github.io/aria-practices/examples/tabs/tabs-1/tabs.html) -- every entry, including roles, has a corresponding element. In result, the test would only be as misleading to a developer (about the difference between the example's implementation of the pattern and the pattern itself) as the tables are on the example page.

But your concern does hark back to the question of test flexibility vs rigidity. Maybe we should be more flexible here. The examples typically won't break just because element of the role changes, so perhaps it is not an interesting or necessary test for most examples.

Do you have a final recommendation here, @tatermelon? Maybe @mcking65 should weigh in as well. Should I remove tests of the role's element from all previous tests and instead only verify that the role exists on the expected number of elements?

mcking65 replied:
Unsure ... simplicity is assert only what is necessary.

One thing we may want to do is change the wording of the message so it is more like "role tab is expected on element button in this implementation." instead of saying that it should only be found on role button, which is not true.

In some cases we may have specific dependencies on html elements in the example. For instance, if the tablist were on a ul, then we would need role presentation on child LI elements. But, that is not necessary if you have a div with role tablist and that div contains buttons. We have some dependencies like that in the menu examples. Also, if you used a div to make tab elements instead of buttons, then the div that is in the page Tab sequence would require tabindex of 0, but if you use a button element, it does not.

What's most important is that the tab elements are descendants of the tablist element. We have structural requirements like that in all the examples. Typically, changing the specific html element used will not change the structural releationships in the DOM.

We want to lean toward simplicity, even if that adds brittleness. At the same time, we do not want to make unnecessary assertions. We also do not want failure messages that imply a requirement that does not exist.

spectranaut commented:
I addressed all your feedback in the latest commit @tatermelon, except for the role question. Can we make a subtask if we change the direction for testing roles and merge these test into bocoup, @mcking65 ?
